src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/ConditionalDeleteResourceHandler.cs

```diff
@@ -0,0 +1,124 @@
+
+            if (request.DeleteMultiple)
+            {
+                return await DeleteMultiple(request, cancellationToken);
+            }
+
+            return await DeleteSingle(request, cancellationToken);
+        }
+
+        private async Task<DeleteResourceResponse> DeleteSingle(ConditionalDeleteResourceRequest request, CancellationToken cancellationToken)
+        {
+            IReadOnlyCollection<SearchResultEntry> matchedResults = await _searchService.ConditionalSearchAsync(request.ResourceType, request.ConditionalParameters, cancellationToken);
+
+            int count = matchedResults.Count;
+            if (count == 0)
+            {
+                return null;
+            }
+            else if (count == 1)
+            {
+                return await _mediator.Send(new DeleteResourceRequest(request.ResourceType, matchedResults.First().Resource.ResourceId, request.HardDelete), cancellationToken);
+            }
+            else
+            {
+                // Multiple matches: The server returns a 412 Precondition Failed error indicating the client's criteria were not selective enough
+                throw new PreconditionFailedException(Core.Resources.ConditionalOperationNotSelectiveEnough);
+            }
+        }
+
+        private async Task<DeleteResourceResponse> DeleteMultiple(ConditionalDeleteResourceRequest request, CancellationToken cancellationToken)
+        {
+            (IReadOnlyCollection<SearchResultEntry> matchedResults, string ct) = await _searchService.ConditionalSearchAsync(request.ResourceType, request.ConditionalParameters, _conditionalDeleteMaxItems, cancellationToken);
+
+            int itemsDeleted = 0;
+
+            // Delete the matched results...
+            while (matchedResults.Any() || !string.IsNullOrEmpty(ct))
+            {
+                foreach (IEnumerable<SearchResultEntry> batch in matchedResults.Take(_conditionalDeleteMaxItems - itemsDeleted).TakeBatch(10))
+                {
+                    DeleteResourceResponse[] results = await Task.WhenAll(batch.Select(result => _mediator.Send(new DeleteResourceRequest(request.ResourceType, result.Resource.ResourceId, request.HardDelete), cancellationToken)));
+                    itemsDeleted += results.Sum(x => x.ResourcesDeleted);
+                }
+
+                if (!string.IsNullOrEmpty(ct) && _conditionalDeleteMaxItems - itemsDeleted > 0)
+                {
+                    (matchedResults, ct) = await _searchService.ConditionalSearchAsync(
+                        request.ResourceType,
+                        request.ConditionalParameters,
+                        _conditionalDeleteMaxItems - itemsDeleted,
+                        cancellationToken,
+                        ct);
+                }
+                else
+                {
+                    matchedResults = Array.Empty<SearchResultEntry>();
+                }
+            }
+
+            return new DeleteResourceResponse(itemsDeleted);
```

Comment:
>itemsDeleted [](http://example.com/codeflow?start=46&length=12)

If itemsDeleted ==0  I think it should return null instead of response.

Also FhirController should properly handle null as response and return Ok as we do in other conditional cases... Not sure about that one, tho

Reply:
Changed the other return to have 0, I think I like having the outcome explicit